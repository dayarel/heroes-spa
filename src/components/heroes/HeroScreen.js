import React from "react";
import { useParams, Redirect } from "react-router-dom";
import { getHeroById } from "../../selectors/getHeroById";
const HeroScreen = ({ history }) => {
  const { heroeid } = useParams();

  const hero = getHeroById(heroeid);

  if (!hero) {
    return <Redirect to="/" />;
  }

  const handleReturn = () => {
    if (history.length <= 2) {
      history.push("/");
    } else {
      history.goBack();
    }
  };

  const {
    id,
    superhero,
    alter_ego,
    first_appearance,
    characters,
    publisher,
  } = hero;

  return (
    <div>
      <div className="row mt-5">
        <div className="col-4">
          <img
            src={`../assets/heroes/${id}.jpg`}
            className="img-thumbnail"
            alt={superhero}
          />
        </div>
        <div className="col-8">
          <h3>{superhero}</h3>
          <ul className="list-group list-group-flush">
            <li className="list-group-item">
              <b>Alter ego: </b>
              {alter_ego}
            </li>
            <li className="list-group-item">
              <b>Publisher: </b>
              {publisher}
            </li>
            <li className="list-group-item">
              <b>First appearance: </b>
              {first_appearance}
            </li>
          </ul>

          <h5>Characters</h5>
          <p>{characters}</p>
          <button className="btn btn-outline-info" onClick={handleReturn}>
            Return
          </button>
        </div>
      </div>
    </div>
  );
};

export default HeroScreen;
