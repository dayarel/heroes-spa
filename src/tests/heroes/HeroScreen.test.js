import React from "react";
import { mount } from "enzyme";
import HeroScreen from "../../components/heroes/HeroScreen";
import { MemoryRouter, Route } from "react-router-dom";
describe("Tests on <HeroScreen />", () => {
  test("should show the component redirect if there's no arguments in the url", () => {
    const historyMock = {
      length: 10,
      push: jest.fn(),
      goBack: jest.fn(),
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("Redirect").exists()).toBe(true);
  });

  test("should show a hero if the url param exists", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-martian"]}>
        <Route path="/hero/:heroeid" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find(".row").exists()).toBe(true);
  });

  test("should go to initial page with push", () => {
    const historyMock = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-martian"]}>
        <Route
          path="/hero/:heroeid"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(historyMock.push).toHaveBeenCalled();
    expect(historyMock.goBack).not.toHaveBeenCalled();
  });

  test("should go back to previous page with push", () => {
    const historyMock = {
      length: 3,
      push: jest.fn(),
      goBack: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-martian"]}>
        <Route
          path="/hero/:heroeid"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(historyMock.goBack).toHaveBeenCalled();
    expect(historyMock.push).not.toHaveBeenCalled();
  });

  test("should call redirect if heroe does not exist", () => {
    const historyMock = {
      length: 3,
      push: jest.fn(),
      goBack: jest.fn(),
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-martian1234567"]}>
        <Route
          path="/hero/:heroeid"
          component={(props) => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    expect(wrapper.text()).toBe("");
  });
});
