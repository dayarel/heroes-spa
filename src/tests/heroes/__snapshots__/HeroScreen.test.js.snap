// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <HeroScreen /> should show the component redirect if there's no arguments in the url 1`] = `""`;
